Update partial-classes-and-methods.md (#39911)
dotnet/docs: docs/csharp/programming-guide/classes-and-structs/partial-classes-and-methods.md

---
title: "Partial Classes and Methods - C# Programming Guide"
description: Partial classes and methods in C# split the definition of a class, a struct, an interface, or a method over two or more source files.
ms.date: 06/21/2021
ms.date: 03/12/2024
helpviewer_keywords:
  - "partial methods [C#]"
  - "partial classes [C#]"
  - "C# language, partial classes and methods"

> ms.assetid: 804cecb7-62db-4f97-a99f-60975bd59fa1

Reviewer: <3

> It's possible to split the definition of a [class](../../language-reference/keywords/class.md), a [struct](../../language-reference/builtin-types/struct.md), an [interface](../../language-reference/keywords/interface.md) or a method over two or more source files. Each source file contains a section of the type or method definition, and all parts are combined when the application is compiled.

Reviewer: ```suggestion
It's possible to split the definition of a [class](../../language-reference/keywords/class.md), a [struct](../../language-reference/builtin-types/struct.md), an [interface](../../language-reference/keywords/interface.md), or a method over two or more source files. Each source file contains a section of the type or method definition, and all parts are combined when the application is compiled.
```

## Partial Classes

There are several situations when splitting a class definition is desirable:

- Declaring a class over separate files enables multiple programmers to work on it at the same time.

> - Code can be added to the class without having to recreate the source file that includes automatically generated source. Visual Studio uses this approach when it creates Windows Forms, Web service wrapper code, and so on. You can create code that uses these classes without having to modify the file created by Visual Studio.

Reviewer: ```suggestion
- You can add code to the class without having to recreate the source file that includes automatically generated source. Visual Studio uses this approach when it creates Windows Forms, Web service wrapper code, and so on. You can create code that uses these classes without having to modify the file created by Visual Studio.
```

For example, consider the following declarations:

:::code language="csharp" source="snippets/partial-classes-and-methods/Program.cs" id="Snippet5":::

They're equivalent to the following declarations:

:::code language="csharp" source="snippets/partial-classes-and-methods/Program.cs" id="Snippet6":::

### Restrictions

There are several rules to follow when you're working with partial class definitions:

- All partial-type definitions meant to be parts of the same type must be modified with `partial`. For example, the following class declarations generate an error:
  :::code language="csharp" source="snippets/partial-classes-and-methods/Program.cs" id="Snippet7":::

> - The `partial` modifier can only appear immediately before the keywords `class`, `struct`, or `interface`.

Reviewer: ```suggestion
- The `partial` modifier can only appear immediately before the keyword `class`, `struct`, or `interface`.
```

:::code language="csharp" source="snippets/partial-classes-and-methods/Program.cs" id="Snippet9":::

The following example shows that you can also develop partial structs and interfaces.

:::code language="csharp" source="snippets/partial-classes-and-methods/Program.cs" id="Snippet10":::

## Partial Methods

A partial class or struct can contain a partial method. One part of the class contains the signature of the method. An implementation can be defined in the same part or another part.

An implementation isn't required for a partial method when the signature obeys the following rules:

- The declaration doesn't include any access modifiers. The method has [`private`](../../language-reference/keywords/private.md) access by default.
- The return type is [`void`](../../language-reference/builtin-types/void.md).
- None of the parameters have the [`out`](../../language-reference/keywords/method-parameters.md#out-parameter-modifier) modifier.
- The method declaration can't include any of the following modifiers:

>   - [virtual](../../language-reference/keywords/virtual.md),

Reviewer: I don't think these bullets should have commas or periods.

>   - [virtual](../../language-reference/keywords/virtual.md),
>   - [override](../../language-reference/keywords/override.md),
>   - [sealed](../../language-reference/keywords/sealed.md),
>   - [new](../../language-reference/keywords/new-modifier.md),
>   - [extern](../../language-reference/keywords/extern.md).

Author: ```suggestion
  - [virtual](../../language-reference/keywords/virtual.md)
  - [override](../../language-reference/keywords/override.md)
  - [sealed](../../language-reference/keywords/sealed.md)
  - [new](../../language-reference/keywords/new-modifier.md)
  - [extern](../../language-reference/keywords/extern.md)
```
